shanghai/core_plugins/channel.py

```diff
@@ -0,0 +1,305 @@
+        del ctx.users[lnick]
+
+
+# Manipulation API
+def remove_nick_from_channel(ctx: NetworkContext, nick, channel):
+    lnick = ctx.nick_lower(nick)
+    lchannel = ctx.nick_lower(channel)
+
+    if ctx.nick_cmp(ctx.network.nickname, nick):
+        # remove myself -> remove channel instance and all users, and remove
+        # all user instances that are not visible anymore.
+
+        # find all joins for that channel, remove them, and collect the nicks
+        # that where on that channel.
+        nicks_to_test = set()
+        for lkey in list(ctx.joins):
+            if ctx.chan_cmp(lchannel, lkey[0]):
+                if not ctx.nick_cmp(ctx.network.nickname, lkey[1]):
+                    nicks_to_test.add(lkey[1])
+                del ctx.joins[lkey]
+
+        # check if nicks that got removed are visible somewhere else
+        for other_nick in list(nicks_to_test):
+            for lkey in ctx.joins:
+                if ctx.nick_cmp(other_nick, lkey[1]):
+                    # nick is visible, remove it from the set
+                    nicks_to_test.remove(other_nick)
+                    break
+
+        # remove invisible nicks
+        for other_nick in nicks_to_test:
+            lother_nick = ctx.nick_lower(other_nick)
+            if lother_nick in ctx.users:
+                del ctx.users[lother_nick]
+
+        # finally, remove the channel
+        if lchannel in ctx.channels:
+            del ctx.channels[lchannel]
+        return
+
+    # case: nick that is not me is removed
+    # remove join, and test if user is still visible. if not, remove user
+    if (lchannel, lnick) in ctx.joins:
+        del ctx.joins[(lchannel, lnick)]
+
+    for lkey in ctx.joins:
+        if ctx.nick_cmp(lnick, lkey[1]):
+            break
+    else:
+        # nick not found
+        if lnick in ctx.users:
+            del ctx.users[lnick]
+
+
+# nick/chan string API
+def nick_lower(ctx: NetworkContext, nick: str):
+    case_mapping = ctx.network.options.get('CASEMAPPING', 'rfc1459').lower()
+    if case_mapping not in ('ascii', 'rfc1459', 'strict-rfc1459'):
+        case_mapping = 'rfc1459'
+    upper_str = string.ascii_uppercase
+    lower_str = string.ascii_lowercase
+    if case_mapping == 'rfc1459':
+        upper_str += '[]\\^'
+        lower_str += '{}|~'
+    elif case_mapping == 'strict-rfc1459':
+        upper_str += '[]\\'
+        lower_str += '{}|'
+    table = nick.maketrans(upper_str, lower_str)
+    return nick.translate(table)
+
+
+def chan_lower(ctx: NetworkContext, chan: str):
+    return ctx.nick_lower(chan)
+
+
+def nick_cmp(ctx: NetworkContext, nick1: str, nick2: str):
+    return ctx.nick_lower(nick1) == ctx.nick_lower(nick2)
+
+
+def chan_cmp(ctx: NetworkContext, chan1: str, chan2: str):
+    return ctx.chan_lower(chan1) == ctx.chan_lower(chan2)
+
+
+@global_event(GlobalEventName.INIT_NETWORK_CTX)
+async def init_context(ctx: NetworkContext):
+    ctx.add_attribute('channels', {})  # l(channel-name) -> channel
```

Comment:
Wasn't meant as type annotation, but rather a simple not, what value is expected for the key (in this case a lowercase version of the channel name)

I'm still kinda unsure about the dict thing. I thought it might as well work to just use `__hash__` and make it return the lowercase version of the channel name and then just use a `set`. Same for users and possibly joins.